concordium-consensus/src/Concordium/KonsensusV1/Consensus.hs

```diff
@@ -0,0 +1,225 @@
+newtype BakerContext = BakerContext
+    { _bakerIdentity :: BakerIdentity
+    }
+
+makeClassy ''BakerContext
+
+-- |Adds a transaction into the pending transaction table
+-- if it's eligible.
+--
+-- Pre condition: A transaction must've been pre-verified prior to being called here.
+--
+-- Transactions received individually are always added to the pending transactions as
+-- it is checked that the transaction nonce is at least what is recorded for the focus block.
+-- (That is a pre condition of this function)
+--
+-- This ensures the invariant of the pending transaction table and the focus block.
+-- Namely that the recorded next availble nonce with respect to the pending transaction table
+-- must be the same of what is recorded in the focus block.
+--
+-- For transactions received as part of a block we must check that the transaction nonce
+-- is at least what the next available nonce recorded in the focus block before adding
+-- it to the pending transaction table.
+--
+-- This is to ensure the above mentioned invariant of the pending transaction table and focus block
+-- as to make sure that if the parent block we verified the transaction within is above the focus block
+-- then we need to record this fact in the pending transaction table as the transaction nonce
+-- would very likely be above what is recorded in the focus block.
+--
+-- This is an internal function only and should not be called directly.
+addPendingTransaction ::
+    ( MonadState (SkovData (MPV m)) m,
+      TimeMonad m,
+      BlockStateQuery m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState (MPV m)
+    ) =>
+    -- |Origin of the transaction.
+    TVer.TransactionOrigin ->
+    -- |The transaction.
+    BlockItem ->
+    m ()
+addPendingTransaction origin bi = do
+    case wmdData bi of
+        NormalTransaction tx -> do
+            fbState <- bpState <$> (_focusBlock <$> gets' _skovPendingTransactions)
+            macct <- getAccount fbState $! transactionSender tx
+            nextNonce <- fromMaybe minNonce <$> mapM (getAccountNonce . snd) macct
+            when (nextNonce <= transactionNonce tx || origin == TVer.Individual) $ do
+                pendingTransactionTable %=! TT.addPendingTransaction nextNonce tx
+                purgeTransactionTable False =<< currentTime
+        CredentialDeployment _ -> do
+            pendingTransactionTable %=! TT.addPendingDeployCredential txHash
+            purgeTransactionTable False =<< currentTime
+        ChainUpdate cu -> do
+            fbState <- bpState <$> (_focusBlock <$> gets' _skovPendingTransactions)
+            nextSN <- getNextUpdateSequenceNumber fbState (updateType (uiPayload cu))
+            when (nextSN <= updateSeqNumber (uiHeader cu) || origin == TVer.Individual) $ do
+                pendingTransactionTable %=! TT.addPendingUpdate nextSN cu
+                purgeTransactionTable False =<< currentTime
+  where
+    txHash = getHash bi
```

Comment:
The issue with the transaction origin still needs to be resolved per prior discussion.